File: testClients/adamTest/AI.py
#-*-python-*-
from BaseAI import BaseAI
from GameObject import *
import random

class AI(BaseAI):
  """The class implementing gameplay logic."""
  @staticmethod
  def username():
    return "The BIRDS of Skyrim"

  @staticmethod
  def password():
    return "53282f00d1a0a60"

  ##This function is called once, before your first turn
  def init(self):
    self.hasSpawned = False
    self.change = -((self.playerID * 2) - 1)
    pass

  ##This function is called once, after your last turn
  def end(self):
    print "This was called"
    pass

  ##This function is called each time it is your turn
  ##Return true to end your turn, return false to ask the server for updated information

  randVars = [0,1,2,3,6]

  def run(self):
    if not self.hasSpawned and random.randint(0, 99) <= 10:
      self.hasSpawned = True
    meh = self.randVars[random.randint(0,len(self.randVars) - 1)]
    if self.players[self.playerID].scrapAmount >= 90:
      if self.hasSpawned:
        bleh = 0
        while bleh < 10:
          bleh += 1
          xDROPU = random.randint(0, 1)
          yDROPU = random.randint(0, self.mapHeight - 1)
          if self.playerID == 1:
            xDROPU = self.mapWidth - xDROPU - 1
            yDROPU = self.mapHeight - yDROPU - 1
          if self.players[self.playerID].scrapAmount > self.modelVariants[meh].cost:
            if self.meh(xDROPU, yDROPU) is None:
              self.players[self.playerID].orbitalDrop(xDROPU, yDROPU, meh)
            else:
              self.hasSpawned = True
              bleh = 20
          else:
            bleh = 30
      else:
        self.hasSpawned = False
        for base in self.droids:
          if base.variant == 7 and base.owner != self.playerID:
            if self.meh(base.x + 1, base.y) is None:
              self.players[self.playerID].orbitalDrop(base.x + 1, base.y, 4)
              break
            if self.meh(base.x - 1, base.y) is None:
              self.players[self.playerID].orbitalDrop(base.x - 1, base.y, 4)
              break
            if self.meh(base.x, base.y + 1) is None:
              self.players[self.playerID].orbitalDrop(base.x, base.y + 1, 4)
              break
            if self.meh(base.x, base.y - 1) is None:
              self.players[self.playerID].orbitalDrop(base.x, base.y - 1, 4)
              break

    for droid in self.droids:
      if droid.healthLeft == 0:
        continue
      if ((droid.owner == self.playerID and droid.hackedTurnsLeft <= 0) or\
          (droid.owner != self.playerID and droid.hackedTurnsLeft > 0))\
           and droid.variant != 7 and droid.variant != 5:
        bleh = []
        if droid.variant == 3:
          for droid2 in self.droids:
            if droid2.owner != self.playerID and droid2.variant != 7 and droid2.variant != 5:
              if abs(droid2.x - droid.x) + abs(droid2.y - droid.y) <= droid.range + droid.maxMovement:
                bleh.append(droid2)
        elif droid.attack > 0:
          for droid2 in self.droids:
            if droid2.owner != (self.playerID ^ (droid2.hackedTurnsLeft > 0)):
              if abs(droid2.x - droid.x) + abs(droid2.y - droid.y) <= droid.range + droid.maxMovement:
                bleh.append(droid2)
        else:
          for droid2 in self.droids:
            if droid2.owner == self.playerID and droid2.id != droid.id and (droid2.maxArmor != droid2.armor or droid2.hackets > 0):
              if abs(droid2.x - droid.x) + abs(droid2.y - droid.y) <= droid.range + droid.maxMovement:
                bleh.append(droid2)
          if len(bleh) == 0:
            bleh.append(droid)
        #attack even if a turret
        for droid2 in bleh:
          for _ in range(droid.maxAttacks):
            if self.okay(droid, droid2):
              droid.operate(droid2.x, droid2.y)

        target2 = None
        self.distance = 99999
        for target in self.droids:
          if droid.attack > 0 and droid.variant != 3:
            if target.variant == 7 and target.owner != self.playerID:
              if (abs(target.x - droid.x) + abs(target.y - droid.y)) < self.distance:
                self.distance = (abs(target.x - droid.x) + abs(target.y - droid.y))
                target2 = target
          elif droid.attack < 0:
            if target.owner == self.playerID and target.id != droid.id:
              if (abs(target.x - droid.x) + abs(target.y - droid.y)) < self.distance:
                self.distance = (abs(target.x - droid.x) + abs(target.y - droid.y))
                target2 = target
          else:
            if target.owner != self.playerID and target.hackedTurnsLeft == 0:
              if target.variant != 5 and target.variant != 7 and target.attack > 0:
                if (abs(target.x - droid.x) + abs(target.y - droid.y)) < self.distance:
                  self.distance = (abs(target.x - droid.x) + abs(target.y - droid.y))
                  target2 = target

        target = target2

        movez = droid.maxMovement

        while movez > 0:
          movez -= 1
          movey = 1

          for droid2 in bleh:
            for _ in range(droid.maxAttacks):
              if self.okay(droid, droid2):
                droid.operate(droid2.x, droid2.y)

          move = True
          for droid2 in self.droids:
            if droid2.id != droid.id:
              if abs(droid2.y - droid.y) + abs(droid2.x - droid.x) <= droid.range:
                if droid2.owner != (self.playerID ^ (droid2.hackedTurnsLeft > 0)) and droid.attack > 0:
                  move = False

          if move and droid.movementLeft > 0:
            if target is not None:
              if target.x > droid.x:
                if self.meh(droid.x + 1, droid.y) is None and droid.x != self.mapWidth - 1:
                  droid.move(droid.x + 1, droid.y)
                elif self.meh(droid.x, droid.y - 1) is None and droid.y != 0:
                  droid.move(droid.x, droid.y - 1)
                elif self.meh(droid.x, droid.y + 1) is None and droid.y != self.mapHeight - 1:
                  droid.move(droid.x, droid.y + 1)
              elif target.x < droid.x:
                if self.meh(droid.x - 1, droid.y) is None and droid.x != 0:
                  droid.move(droid.x - 1, droid.y)
                elif self.meh(droid.x, droid.y - 1) is None and droid.y != 0:
                  droid.move(droid.x, droid.y - 1)
                elif self.meh(droid.x, droid.y + 1) is None and droid.y != self.mapHeight - 1:
                  droid.move(droid.x, droid.y + 1)
              elif target.y > droid.y:
                if self.meh(droid.x, droid.y + 1) is None and droid.y != self.mapHeight - 1:
                  droid.move(droid.x, droid.y + 1)
                elif self.meh(droid.x - 1, droid.y) is None and droid.x != 0:
                  droid.move(droid.x - 1, droid.y)
                elif self.meh(droid.x + 1, droid.y) is None and droid.x != self.mapWidth - 1:
                  droid.move(droid.x + 1, droid.y )
              elif target.y < droid.y:
                if self.meh(droid.x, droid.y - 1) is None and droid.y != 0:
                  droid.move(droid.x, droid.y - 1)
                elif self.meh(droid.x - 1, droid.y) is None and droid.x != 0:
                  droid.move(droid.x - 1, droid.y)
                elif self.meh(droid.x + 1, droid.y) is None and droid.x != self.mapWidth - 1:
                  droid.move(droid.x + 1, droid.y )
    return 1

  def meh(self, x, y):
    for droid in self.droids:
      if x == droid.x and y == droid.y:
        return droid
    return None

  def dist(self, me, you):
    return abs(me.x  - you.x) + abs(me.y - you.y)

  def okay(self, me, target):
    if me.attacksLeft == 0 or target.healthLeft == 0:
      return False
    if me.variant == 3:
      if target.owner != self.playerID and target.hackedTurnsLeft == 0 and self.dist(me, target) <= me.range:
        return True
    elif me.attack < 0:
      if target.owner == self.playerID and self.dist(me, target) <= me.range and (target.maxArmor != target.armor or target.hackets > 0):
        return True
    else:
      if target.owner != (self.playerID ^ (target.hackedTurnsLeft > 0)) and self.dist(me, target) <= me.range:
        return True
    return False

  def __init__(self, conn):
    BaseAI.__init__(self, conn)

File: ShellAI/cpp/AI.cpp
#include "AI.h"
#include "util.h"

enum
{
  CLAW = 0,
  ARCHER = 1,
  REPAIRER = 2,
  HACKER = 3,
  TURRET = 4,
  WALL = 5,
  TERMINATOR = 6,
  HANGAR = 7,
};

AI::AI(Connection* conn) : BaseAI(conn) {}

const char* AI::username()
{
  return "Shell AI";
}

const char* AI::password()
{
  return "password";
}

//This function is run once, before your first turn.
void AI::init()
{
  int offset = 0;
  bool found = false;
  while(!found)
  {
    //find a location without a hangar
    for(int i = 0; i < tiles.size(); i++)
    {
      //make sure that the tile is near the edge
      if(tiles[i].x() == (mapWidth() - 1) * playerID() + offset)
      {
        bool hangarPresent = false;
        //check for hangar
        for(int z = 0; z < droids.size(); z++)
        {
          if(droids[z].x() == tiles[i].x() && droids[z].y() == tiles[i].y())
          {
            hangarPresent = true;
            break;
          }
        }
        if(!hangarPresent)
        {
          spawnX = tiles[i].x();
          spawnY = tiles[i].y();
          found = true;
          break;
        }
      }
    }
    //if nothing was found then move away from the edge
    if(!found)
    {
      //if on the left
      if(playerID() == 0)
      {
        offset++;
      }
      else
      {
        //on the right
        offset--;
      }
    }
  }
}

//This function is called each time it is your turn.
//Return true to end your turn, return false to ask the server for updated information.
bool AI::run()
{
  //try to spawn a claw near your side
  //make sure you own enough scrap
  if(players[playerID()].scrapAmount() >= modelVariants[CLAW].cost())
  {
    //make sure nothing is spawning there
    if(getTile(spawnX, spawnY)->turnsUntilAssembled() == 0)
    {
      bool spawn = true;
      //make sure there isn't a hangar there
      for(int i = 0; i < droids.size(); i++)
      {
        //if the droid's x and y is the same as the spawn point
        if(droids[i].x() == spawnX && droids[i].y() == spawnY)
        {
          //if the droid is a hangar
          if(droids[i].variant() == HANGAR)
          {
            //can't spawn on top of hangars
            spawn = false;
            break;
          }
        }
      }
      if(spawn)
      {
        //spawn the claw
        players[playerID()].orbitalDrop(spawnX, spawnY, CLAW);
      }
    }
  }
  //loop through all of the droids
  for(int i = 0; i < droids.size(); i++)
  {
    //if you have control of the droid
    if((droids[i].owner() == playerID() && droids[i].hackedTurnsLeft() <= 0) ||
       (droids[i].owner() != playerID() && droids[i].hackedTurnsLeft() > 0))
    {
      //if there are any moves to be done
      if(droids[i].movementLeft() > 0)
      {
        //try to move towards the enemy
        int changeX = 1;
        //if on the right move towards the left
        if(playerID() == 1)
        {
          changeX = -1;
        }
        bool move = true;
        //check if there is a droid on that tile
        for(int z = 0; z < droids.size(); z++)
        {
          //if the two droids are different
          if(droids[z].id() != droids[i].id())
          {
            //if there is a droid to run into
            if(droids[z].x() == droids[i].x() + changeX && droids[z].y() == droids[i].y())
            {
              //don't move
              move = false;
            }
          }
        }
        //move if okay and within map boundaries
        if(move && droids[i].x() + changeX >= 0 && droids[i].x() + changeX < mapWidth())
        {
          droids[i].move(droids[i].x() + changeX, droids[i].y());
        }
      }
      //if there are any attacks left
      if(droids[i].attacksLeft() > 0)
      {
        //find a target towards the enemy
        int changeX = 1;
        //enemy is to the left if playerID is one
        if(playerID() == 1)
        {
          changeX = -1;
        }
        Droid* target = NULL;
        for(int z = 0; z < droids.size(); z++)
        {
          //if the droid is there make it a target
          if(droids[z].x() == droids[i].x() + changeX && droids[z].y() == droids[i].y() && droids[z].healthLeft() > 0)
          {
            target = &droids[z];
          }
        }
        //if a target was found
        if(target != NULL)
        {
          //repairer logic
          if(droids[i].variant() == REPAIRER)
          {
            //only try to heal your units or hacked enemy units
            if((target->owner() == playerID() && target->hackedTurnsLeft() <= 0) ||
               (target->owner() != playerID() && target->hackedTurnsLeft() > 0))
            {
              //heal the target
              droids[i].operate(target->x(), target->y());
            }
          }
          //hacker unit logic
          else if(droids[i].variant() == HACKER)
          {
            //only operate on non-hacked enemy units
            if(target->owner() != playerID() && target->hackedTurnsLeft() > 0)
            {
              //don't hack hangars or walls
              if(target->variant() != HANGAR && target->variant() != WALL)
              {
                //hack the target
                droids[i].operate(target->x(), target->y());
              }
            }
          }
          //other unit logic
          else
          {
            //only operate on hacked friendly units or enemy units
            if((target->owner() == playerID() && target->hackedTurnsLeft() > 0) ||
               (target->owner() != playerID() && target->hackedTurnsLeft() <= 0))
            {
              //attack the target
              droids[i].operate(target->x(), target->y());
            }
          }
        }
      }
    }
  }
  return true;
}

//This function is run once, after your last turn.
void AI::end(){}

//This functions returns a pointer to a tile, or returns NULL for an invalid tile
Tile* AI::getTile(int x, int y)
{
  if(x >= mapWidth() || x < 0 || y >= mapHeight() || y < 0)
  {
    return NULL;
  }
  return &tiles[y + x * mapHeight()];
}
